x-pack/test/functional/services/ml/data_frame_analytics_edit.ts

```diff
@@ -7,17 +7,206 @@
+      await this.clickOpenCustomUrlEditor();
+      await customUrls.setCustomUrlLabel(customUrl.label);
+      await mlCommonUI.selectRadioGroupValue('mlJobCustomUrlLinkToTypeInput', URL_TYPE.OTHER);
+      await customUrls.setCustomUrlOtherTypeUrl(customUrl.url);
+    },
+
+    async addDashboardCustomUrl(
+      customUrl: DashboardUrlConfig,
+      expectedResult: { index: number; url: string }
+    ) {
+      await retry.tryForTime(30 * 1000, async () => {
+        await this.fillInDashboardUrlForm(customUrl);
+      });
+      await this.saveCustomUrl(customUrl.label, expectedResult.index, expectedResult.url);
+    },
+
+    async addDiscoverCustomUrl(customUrl: DiscoverUrlConfig) {
+      await retry.tryForTime(30 * 1000, async () => {
+        const existingCustomUrlCount = await this.getExistingCustomUrlCount();
+
+        await this.fillInDiscoverUrlForm(customUrl);
+        await this.saveCustomUrl(customUrl.label, existingCustomUrlCount);
+      });
+    },
+
+    async addOtherTypeCustomUrl(customUrl: OtherUrlConfig) {
+      await retry.tryForTime(30 * 1000, async () => {
+        const existingCustomUrlCount = await this.getExistingCustomUrlCount();
+
+        await this.fillInOtherUrlForm(customUrl);
+        await this.saveCustomUrl(customUrl.label, existingCustomUrlCount);
+      });
+    },
+
+    async testOtherTypeCustomUrlAction(indexInList: number, expectedUrl: string) {
+      await customUrls.assertCustomUrlUrlValue(indexInList, expectedUrl);
+      await this.closeEditJobFlyout();
+    },
+
+    async editCustomUrl(indexInList: number, customUrl: { label: string; url: string }) {
+      await customUrls.editCustomUrlLabel(indexInList, customUrl.label);
+      await customUrls.editCustomUrlUrlValue(indexInList, customUrl.url);
+    },
+
+    async deleteCustomUrl(jobId: string, indexInList: number) {
+      const beforeCustomUrls = await testSubjects.findAll('mlJobEditCustomUrlItemLabel');
+      await customUrls.deleteCustomUrl(indexInList);
+      return beforeCustomUrls.length;
+    },
+
+    async assertCustomUrlsLength(expectedLength: number) {
+      await customUrls.assertCustomUrlsLength(expectedLength);
+      await this.closeEditJobFlyout();
+    },
+
+    async openTestCustomUrl(jobId: string, indexInList: number) {
+      await this.openEditCustomUrlsForJobTab(jobId);
+      await customUrls.clickTestCustomUrl(indexInList);
+    },
+
+    async testDiscoverCustomUrlAction(expectedHitCountFormatted?: string) {
+      await customUrls.assertDiscoverCustomUrlAction(expectedHitCountFormatted);
+    },
+
+    async testDashboardCustomUrlAction(expectedPanelCount: number) {
+      await customUrls.assertDashboardCustomUrlAction(expectedPanelCount);
+    },
+
+    async closeEditJobFlyout() {
+      if (await testSubjects.exists('mlAnalyticsEditFlyoutCancelButton')) {
+        await testSubjects.click('mlAnalyticsEditFlyoutCancelButton');
+        await testSubjects.missingOrFail('mlAnalyticsEditFlyout');
```

Comment:
Suggestion to use `this.assertAnalyticsEditFlyoutMissing()` instead and to add retry in case of flaky CI.